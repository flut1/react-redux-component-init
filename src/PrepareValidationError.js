import ExtendableError from './utils/ExtendableError';

class PrepareValidationError extends ExtendableError {}

export default PrepareValidationError;
