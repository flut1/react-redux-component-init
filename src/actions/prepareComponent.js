import { MODE_PREPARE } from '../initMode';
import extractValuesForProps from '../utils/extractValuesForProps';

import initComponent from './initComponent';

/**
 * Action that prepares a single component. Duplicate calls with exactly the same
 * `Component` and `props` will be ignored.
 *
 * @param {react.Component} Component The component to prepare. If the component was not
 * wrapped in `withInitAction`, this action will have no effect and returns a Promise that
 * resolves immediately.
 * @param {object} [props={}] The values of `initProps` to initialize this component with. These
 * values should be the same as the props that the component will eventually mount with. Omitting
 * one of the `initProps` values configured in `withInitAction()` will result in an error.
 * @returns {function} A thunk function that should be passed directly to the Redux `dispatch`
 * function.
 * @example dispatch(prepareComponent(Comment, { commentId: 5 }));
 */
export default (Component, props = {}) =>
  (dispatch, getState) => {
    // unwrap other HoC if no initConfig is found
    while (!Component.initConfig && Component.WrappedComponent) {
      Component = Component.WrappedComponent; // eslint-disable-line no-param-reassign
    }

    if (Component.initConfig) {
      const {
        componentId,
        initProps,
        options: { getInitState, getPrepareKey },
      } = Component.initConfig;

      const initState = getInitState(getState());
      if (!initState) {
        throw new ReferenceError('Could not find init state. Did you attach the init reducer?');
      }

      const { mode } = initState;
      if (mode === MODE_PREPARE) {
        const initValues = extractValuesForProps(props, initProps);

        initValues.forEach((initValue, index) => {
          if (typeof initValue === 'undefined') {
            throw new ReferenceError(`Component "${componentId}" expected prop "${initProps[index]}" but was not passed to prepareComponent`);
          }
        });

        const prepareKey = getPrepareKey(componentId, initValues);

        return dispatch(initComponent(Component, initValues, prepareKey, {
          isPrepare: true,
        }));
      }
    }

    return Promise.resolve();
  };
