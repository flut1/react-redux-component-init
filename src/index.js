export { default as withInitAction } from './withInitAction';
export { default as initReducer } from './reducer';

export { default as initComponent } from './actions/initComponent';
export { default as prepareComponent } from './actions/prepareComponent';
export { default as prepareComponents } from './actions/prepareComponents';
export { default as setInitMode } from './actions/setInitMode';

export { default as PrepareValidationError } from './PrepareValidationError';
export * from './initMode';
export * from './initSelfMode';
export * as actionTypes from './actions/actionTypes';
