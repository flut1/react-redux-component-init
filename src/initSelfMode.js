/**
 * Possible values for the `selfInit` option of `withInitAction()`
 */

/**
 * The component will render immediately, even if `initAction` is still pending.
 * It is recommended to use this option and render a loading indicator or placeholder content
 * until `initAction` is resolved. This will give the user immediate feedback that something is
 * being loaded. While the `initAction` is pending, an `isInitializing` prop will be passed to
 * the component.
 * @type {string}
 */
export const INIT_SELF_ASYNC = 'ASYNC';

/**
 * This will cause the component not to be mounted until the first initialization has completed.
 * The component will remain mounted during any reinitialization.
 * @type {string}
 */
export const INIT_SELF_BLOCKING = 'BLOCKING';

/**
 * The same behavior as `INIT_SELF_BLOCKING` but it will also unmount the component during
 * reinitialization.
 * @type {string}
 */
export const INIT_SELF_UNMOUNT = 'UNMOUNT';

/**
 * Only execute initAction on the server (initMode == MODE_PREPARE). Initialization will
 * be skipped on the client.
 * @type {string}
 */
export const INIT_SELF_NEVER = 'NEVER';
