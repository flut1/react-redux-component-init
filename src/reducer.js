import { createSelector } from 'reselect';

import createShallowArrayCompareSelector from './utils/createShallowArrayCompareSelector';
import { INIT_COMPONENT, SET_INIT_MODE } from './actions/actionTypes';
import extractValuesForProps from './utils/extractValuesForProps';
import { MODE_PREPARE } from './initMode';

/**
 * Reducer function that manages the state for `react-redux-component-init`. This reducer should
 * be included using Redux's `combineReducers()` under the `init` key in the root of
 * the store state.
 */
export default (state = {
  mode: MODE_PREPARE,
  prepared: {},
  selfInit: {},
}, action) => {
  switch (action.type) {
    case SET_INIT_MODE:
      return {
        ...state,
        mode: action.payload,
      };
    case INIT_COMPONENT:
      if (action.payload.isPrepare) {
        return {
          ...state,
          prepared: {
            ...state.prepared,
            [action.payload.prepareKey]: action.payload.complete,
          },
        };
      }
      return {
        ...state,
        selfInit: {
          ...state.selfInit,
          [action.payload.prepareKey]: action.payload.complete,
        },
      };
    default:
      return state;
  }
};

const createComponentInitValuesSelector = ({
  componentId, initProps, options: { getPrepareKey },
}) =>
  createShallowArrayCompareSelector(
    (state, props) => extractValuesForProps(props, initProps),
    initValues => ({
      prepareKey: getPrepareKey(componentId, initValues),
      initValues,
    }),
  );

export const createComponentInitStateSelector = initConfig => createSelector(
  createComponentInitValuesSelector(initConfig),
  state => state.selfInit,
  state => state.prepared,
  ({ prepareKey, initValues }, selfInit, prepared) => ({
    prepareKey,
    initValues,
    selfInitState: selfInit[prepareKey],
    isPrepared: !!prepared[prepareKey],
  }),
);
