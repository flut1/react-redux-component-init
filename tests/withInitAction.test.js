/* global expect, describe, it, jest */
import React from 'react';

import withInitAction, { clearComponentIds } from '../src/withInitAction';
import initReducer from '../src/reducer';
import prepareComponent from '../src/actions/prepareComponent';
import PrepareValidationError from '../src/PrepareValidationError';
import setInitMode from '../src/actions/setInitMode';
import { MODE_INIT_SELF } from '../src/initMode';

import SimpleInitTestComponent from './fixtures/SimpleInitTestComponent';
import IsomorphicTestEnvironment from './IsomorphicTestEnvironment';

describe('isomorphic application', () => {
  describe('server-rendering a component with prepared initAction', () => {
    describe('and calling prepareComponent() on the server', () => {
      clearComponentIds();
      const mockAction = jest.fn(() => Promise.resolve());
      const TestComponent = withInitAction(mockAction)(SimpleInitTestComponent);

      const environment = new IsomorphicTestEnvironment(() => <TestComponent />, {
        init: initReducer,
      });

      const preparePromise = environment.server.store.dispatch(prepareComponent(TestComponent));

      it('should render the component on the server', async () => {
        await preparePromise;
        const { serverTestRenderer } = environment.server.render();
        expect(serverTestRenderer.toJSON()).toMatchSnapshot();
      });

      it('should render the component on the client', () => {
        const { clientTestRenderer } = environment.client.render();
        expect(clientTestRenderer.toJSON()).toMatchSnapshot();
      });

      it('should call the initAction once', () => {
        expect(mockAction).toHaveBeenCalledTimes(1);
      });
    });

    describe('without calling prepareComponent() on the server', () => {
      clearComponentIds();
      const mockAction = () => Promise.resolve();
      const TestComponent = withInitAction({ prepared: mockAction })(SimpleInitTestComponent);

      const environment = new IsomorphicTestEnvironment(() => <TestComponent />, {
        init: initReducer,
      });

      it('should throw an error on server', () => {
        expect(() => environment.server.render()).toThrow(PrepareValidationError);
      });
    });
  });

  describe('rendering a component with prepared initAction after first-render on client', () => {
    clearComponentIds();
    const mockAction = jest.fn(() => Promise.resolve());

    const TestComponent = withInitAction(mockAction)(SimpleInitTestComponent);

    const environment = new IsomorphicTestEnvironment(() => <div>no TestComponent</div>, {
      init: initReducer,
    });

    environment.server.render();
    const { clientTestRenderer } = environment.client.render();
    environment.client.store.dispatch(setInitMode(MODE_INIT_SELF));
    environment.client.update(() => (
      <div>
        <TestComponent />
      </div>
    ));

    it('should render the component on the client', () => {
      expect(clientTestRenderer.toJSON()).toMatchSnapshot();
    });

    it('should call the initAction once', () => {
      expect(mockAction).toHaveBeenCalledTimes(1);
    });
  });

  describe('server-rendering a component with prepared initAction and initProps', () => {
    describe('calling prepareComponent with the correct prop', () => {
      clearComponentIds();
      const mockAction = jest.fn(({ testInitProp }) => Promise.resolve(testInitProp));
      const TestComponent = withInitAction(['testInitProp'], mockAction)(SimpleInitTestComponent);

      const environment = new IsomorphicTestEnvironment(
        () => <TestComponent testInitProp={5} testProp={15} />,
        { init: initReducer },
      );

      const preparePromise = environment.server.store.dispatch(
        prepareComponent(TestComponent, { testInitProp: 5 }),
      );

      it('should render the component on the server', async () => {
        await preparePromise;
        const { serverTestRenderer } = environment.server.render();
        expect(serverTestRenderer.toJSON()).toMatchSnapshot();
      });

      it('should render the component on the client', () => {
        const { clientTestRenderer } = environment.client.render();
        expect(clientTestRenderer.toJSON()).toMatchSnapshot();
      });

      it('should call the initAction once with the correct prop value', async () => {
        expect(mockAction).toHaveBeenCalledTimes(1);
        await expect(mockAction.mock.results[0].value).resolves.toBe(5);
      });

      describe('changing a non-init prop on the client', () => {
        it('should not have called the initAction again', () => {
          environment.client.store.dispatch(setInitMode(MODE_INIT_SELF));
          environment.client.update(() => <TestComponent testInitProp={5} testProp={16} />);

          return new Promise(resolve => {
            process.nextTick(() => {
              expect(mockAction).toHaveBeenCalledTimes(1);
              resolve();
            });
          });
        });
      });

      describe('changing the init prop on the client', () => {
        it('should call the initAction again with the correct value', () => {
          environment.client.update(() => <TestComponent testInitProp={6} testProp={16} />);

          return new Promise(resolve => {
            process.nextTick(async () => {
              expect(mockAction).toHaveBeenCalledTimes(2);
              await expect(mockAction.mock.results[1].value).resolves.toBe(6);
              resolve();
            });
          });
        });
      });
    });
  });

  describe('server-rendering a component with prepared and clientOnly initAction', () => {
    clearComponentIds();
    const mockActionPrepared = jest.fn(({ testInitProp }) => Promise.resolve(testInitProp));
    const mockActionClient = jest.fn(({ testInitProp }) => Promise.resolve(testInitProp));
    const TestComponent = withInitAction(['testInitProp'], {
      prepared: mockActionPrepared,
      clientOnly: mockActionClient,
    })(SimpleInitTestComponent);

    const environment = new IsomorphicTestEnvironment(
      () => <TestComponent testInitProp={5} testProp={15} />,
      { init: initReducer },
    );

    const preparePromise = environment.server.store.dispatch(
      prepareComponent(TestComponent, { testInitProp: 5 }),
    );

    it('renders the component on the server', async () => {
      await preparePromise;
      const { serverTestRenderer } = environment.server.render();
      expect(serverTestRenderer.toJSON()).toMatchSnapshot();
    });

    it('calls the prepared initAction once with the correct prop value', async () => {
      expect(mockActionPrepared).toHaveBeenCalledTimes(1);
      await expect(mockActionPrepared.mock.results[0].value).resolves.toBe(5);
    });

    it('does not call the clientOnly initAction before client render', () => {
      expect(mockActionClient).not.toHaveBeenCalled();
    });

    it('renders the component on the client', () => {
      const { clientTestRenderer } = environment.client.render();
      expect(clientTestRenderer.toJSON()).toMatchSnapshot();
    });

    it('calls the prepared initAction only once', () => {
      expect(mockActionPrepared).toHaveBeenCalledTimes(1);
    });

    it('calls the client initAction once with the correct prop value', async () => {
      expect(mockActionClient).toHaveBeenCalledTimes(1);
      await expect(mockActionClient.mock.results[0].value).resolves.toBe(5);
    });

    describe('changing a non-init prop on the client', () => {
      it('does not call the initAction again', () => {
        environment.client.store.dispatch(setInitMode(MODE_INIT_SELF));
        environment.client.update(() => <TestComponent testInitProp={5} testProp={16} />);

        return new Promise(resolve => {
          process.nextTick(() => {
            expect(mockActionClient).toHaveBeenCalledTimes(1);
            expect(mockActionPrepared).toHaveBeenCalledTimes(1);
            resolve();
          });
        });
      });
    });

    describe('changing the init prop on the client', () => {
      it('calls both initAction again with the correct value', () => {
        environment.client.update(() => <TestComponent testInitProp={6} testProp={16} />);

        return new Promise(resolve => {
          process.nextTick(async () => {
            expect(mockActionPrepared).toHaveBeenCalledTimes(2);
            await expect(mockActionPrepared.mock.results[1].value).resolves.toBe(6);
            expect(mockActionClient).toHaveBeenCalledTimes(2);
            await expect(mockActionClient.mock.results[1].value).resolves.toBe(6);
            resolve();
          });
        });
      });
    });
  });
});
