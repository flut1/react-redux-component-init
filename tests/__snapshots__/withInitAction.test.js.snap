// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isomorphic application rendering a component with prepared initAction after first-render on client should render the component on the client 1`] = `
<div>
  <div>
    <h1>
      This is a test component
    </h1>
  </div>
</div>
`;

exports[`isomorphic application server-rendering a component with prepared and clientOnly initAction renders the component on the client 1`] = `
<div>
  <h1>
    This is a test component
  </h1>
  <span>
    The test param is: 
    5
  </span>
  <span>
    The test prop is: 
    15
  </span>
</div>
`;

exports[`isomorphic application server-rendering a component with prepared and clientOnly initAction renders the component on the server 1`] = `
<div>
  <h1>
    This is a test component
  </h1>
  <span>
    The test param is: 
    5
  </span>
  <span>
    The test prop is: 
    15
  </span>
</div>
`;

exports[`isomorphic application server-rendering a component with prepared initAction and calling prepareComponent() on the server should render the component on the client 1`] = `
<div>
  <h1>
    This is a test component
  </h1>
</div>
`;

exports[`isomorphic application server-rendering a component with prepared initAction and calling prepareComponent() on the server should render the component on the server 1`] = `
<div>
  <h1>
    This is a test component
  </h1>
</div>
`;

exports[`isomorphic application server-rendering a component with prepared initAction and initProps calling prepareComponent with the correct prop should render the component on the client 1`] = `
<div>
  <h1>
    This is a test component
  </h1>
  <span>
    The test param is: 
    5
  </span>
  <span>
    The test prop is: 
    15
  </span>
</div>
`;

exports[`isomorphic application server-rendering a component with prepared initAction and initProps calling prepareComponent with the correct prop should render the component on the server 1`] = `
<div>
  <h1>
    This is a test component
  </h1>
  <span>
    The test param is: 
    5
  </span>
  <span>
    The test prop is: 
    15
  </span>
</div>
`;
